### ISSUE TITLE
Num of classes of ADE20k, 150 or 151?

### ISSUE THREAD

**Issue by REFunction** [holyseven/PSPNet-TF-Reproduce#25]
Thank you for reproducing pspnet.
I find you set num_classes = 150 instead of 151 while 0 represents 'other' class. I think this may cause a little difference during training and testing.

**Comment by holyseven** [comment]
Hi,
I didn't use the 'other' class for training and test. https://github.com/holyseven/PSPNet-TF-Reproduce/blob/master/database/reader.py#L269-L270.
pspnet used also 150 classes, I think.

**Comment by REFunction** [comment]
> Hi,
> I didn't use the 'other' class for training and test. https://github.com/holyseven/PSPNet-TF-Reproduce/blob/master/database/reader.py#L269-L270.
> pspnet used also 150 classes, I think.

It's so strange.[PSPNet](https://github.com/hszhao/PSPNet/blob/master/evaluation/eval_all.m) set 20+1 for Pascal VOC, 18+1 for Cityscapes while only 150 for ADE20k. Then I check the masks in ADE20k again and find there are definitely some zeros in them.

**Comment by holyseven** [comment]
Usually, the training and test do not involve the background/trash/other class. But it all depends on the creator of the dataset and its evaluation server.
- Pascal VOC, the server computes the accuracy of the background class.
- Cityscapes has 19+1 classes, not 18+1. See https://github.com/holyseven/PSPNet-TF-Reproduce/blob/master/database/helper_cityscapes.py#L61-L98. Note that the trainId starts from 0. and the server computes only 19 classes. 
- As for ADE20K, class 0 means the trash class, which is not involved during training and test (not verified, but quite sure). So it has only 150 classes for training and test.

In my opinion, the difference of Pascal VOC is because its background class occupies a lot of the image.

**Comment by REFunction** [comment]
> Usually, the training and test do not involve the background/trash/other class. But it all depends on the creator of the dataset and its evaluation server.
> 
> * Pascal VOC, the server computes the accuracy of the background class.
> * Cityscapes has 19+1 classes, not 18+1. See https://github.com/holyseven/PSPNet-TF-Reproduce/blob/master/database/helper_cityscapes.py#L61-L98. Note that the trainId starts from 0. and the server computes only 19 classes.
> * As for ADE20K, class 0 means the trash class, which is not involved during training and test (not verified, but quite sure). So it has only 150 classes for training and test.
> 
> In my opinion, the difference of Pascal VOC is because its background class occupies a lot of the image.

Thank you for explaining so much. Now I have got your idea. But  I notice that you only set num_classes=150 and take 'tf.nn.sparse_softmax_cross_entropy_with_logits' as a part of loss function. Can I think it includes class ids from 0 to 149? If so, this will take class 0 into account and ignore class 150, which is inconsistent with expection.

**Comment by holyseven** [comment]
That's why I did this: https://github.com/holyseven/PSPNet-TF-Reproduce/blob/master/database/reader.py#L269-L270, which eliminates the class 0.

**Comment by REFunction** [comment]
> That's why I did this: https://github.com/holyseven/PSPNet-TF-Reproduce/blob/master/database/reader.py#L269-L270, which eliminates the class 0.

Hahahahaha, got it.

**Comment by lakshmankanakala** [comment]
Hi @holyseven, Is there any option to load the last saved checkpoint to train (resume option)

Thank you

**Comment by holyseven** [comment]
@lakshmankanakala Sorry, no... But you can add one.
